drivers/input/input_esp32_touch_sensor.c

```diff
@@ -0,0 +1,335 @@
+	touch_filter_config_t filter_info = {
+		.mode = dev_cfg->filter_mode,
+		.debounce_cnt = dev_cfg->filter_debounce_cnt,
+		.noise_thr = dev_cfg->filter_noise_thr,
+		.jitter_step = dev_cfg->filter_jitter_step,
+		.smh_lvl = dev_cfg->filter_smooth_level,
+	};
+
+	touch_hal_filter_set_config(&filter_info);
+	touch_hal_filter_enable();
+
+	touch_hal_timeout_enable();
+	touch_hal_timeout_set_threshold(SOC_TOUCH_PAD_THRESHOLD_MAX);
+#endif /* defined(CONFIG_SOC_SERIES_ESP32) */
+
+	/* Register touch interrupt ISR, enable intr type. */
+#if defined(CONFIG_SOC_SERIES_ESP32)
+	dev_data->rtc_intr_msk = RTC_CNTL_TOUCH_INT_ST_M;
+	rtc_isr_install(&rtc_isr, (void *)dev);
+	touch_hal_intr_enable();
+#elif defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3)
+	dev_data->rtc_intr_msk = RTC_CNTL_TOUCH_DONE_INT_ST_M
+				|RTC_CNTL_TOUCH_ACTIVE_INT_ST_M
+				|RTC_CNTL_TOUCH_INACTIVE_INT_ST_M
+				|RTC_CNTL_TOUCH_SCAN_DONE_INT_ST_M
+				|RTC_CNTL_TOUCH_TIMEOUT_INT_ST_M;
+	rtc_isr_install(&rtc_isr, (void *)dev);
+	touch_hal_intr_enable(TOUCH_PAD_INTR_MASK_ACTIVE
+			|TOUCH_PAD_INTR_MASK_INACTIVE
+			|TOUCH_PAD_INTR_MASK_TIMEOUT
+			|TOUCH_PAD_INTR_MASK_SCAN_DONE);
+	touch_hal_set_fsm_mode(TOUCH_FSM_MODE_TIMER);
+#endif /* defined(CONFIG_SOC_SERIES_ESP32) */
+
+	touch_hal_start_fsm();
+
+	return 0;
+}
+
+#define TOUCH_SENSOR_CHANNEL_CFG_DEF(node_id)							\
+	{											\
+		.channel_num  = DT_PROP(node_id, channel_num),					\
+		.channel_sens = DT_PROP(node_id, channel_sens),					\
+		.zephyr_code  = DT_PROP(node_id, zephyr_code),					\
+	}
+
+#define TOUCH_SENSOR_INIT(inst)									\
+	static const struct touch_sensor_channel_config						\
+		touch_sensor_channel_config_##inst[] = {					\
+		DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP(inst, TOUCH_SENSOR_CHANNEL_CFG_DEF, (,))};\
+												\
+	static const struct touch_sensor_config touch_sensor_config_##inst = {			\
+		.debounce_interval_ms = DT_INST_PROP(inst, debounce_interval_ms),		\
+		.num_channels = ARRAY_SIZE(touch_sensor_channel_config_##inst),			\
+		.volt_refh = DT_INST_PROP(inst, volt_refh),					\
+		.volt_refl = DT_INST_PROP(inst, volt_refl),					\
+		.volt_atten = DT_INST_PROP(inst, volt_atten),					\
+		.filter_mode = DT_INST_PROP(inst, filter_mode),					\
+		.filter_debounce_cnt = DT_INST_PROP(inst, filter_debounce_cnt),			\
+		.filter_noise_thr = DT_INST_PROP(inst, filter_noise_thr),			\
+		.filter_jitter_step = DT_INST_PROP(inst, filter_jitter_step),			\
+		.filter_smooth_level = DT_INST_PROP(inst, filter_smooth_level),			\
+		.channel_cfg = touch_sensor_channel_config_##inst,				\
+	};											\
+												\
+	static struct touch_sensor_channel_data touch_sensor_channel_data_##inst		\
+				[ARRAY_SIZE(touch_sensor_channel_config_##inst)];		\
+												\
+	static struct touch_sensor_data touch_sensor_data_##inst = {				\
+		.rtc_intr_msk = 0,								\
+		.channel_data = touch_sensor_channel_data_##inst,				\
```

Comment:
I believe `channel_data` must be kept under `data` struct because `touch_sensor_channel_data_##inst` is a vector containing each channel current status that will change during the driver instance life